wasmsdk/player_stream.go

```diff
@@ -0,0 +1,228 @@
+//go:build js && wasm
+// +build js,wasm
+
+package main
+
+import (
+	"context"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/0chain/gosdk/core/sys"
+	"github.com/0chain/gosdk/zboxcore/marker"
+	"github.com/0chain/gosdk/zboxcore/sdk"
+)
+
+type StreamPlayer struct {
+	allocationID string
+	remotePath   string
+	authTicket   string
+	lookupHash   string
+
+	isViewer      bool
+	allocationObj *sdk.Allocation
+	authTicketObj *marker.AuthTicket
+
+	todoQueue   chan sdk.PlaylistFile
+	reloadQueue chan sdk.PlaylistFile
+	latestTodo  sdk.PlaylistFile
+
+	downloadedQueue chan []byte
+	ctx             context.Context
+	cancel          context.CancelFunc
+	prefetchQty     int
+}
+
+func (p *StreamPlayer) Start() error {
+	if p.cancel != nil {
+		p.cancel()
+	}
+
+	p.ctx, p.cancel = context.WithCancel(context.TODO())
+	p.todoQueue = make(chan sdk.PlaylistFile, 100)
+	p.reloadQueue = make(chan sdk.PlaylistFile, p.prefetchQty)
+	p.downloadedQueue = make(chan []byte, p.prefetchQty)
+
+	go p.reloadList()
+	go p.startDownload()
+	go p.nextTodo()
+
+	return nil
+}
+
+func (p *StreamPlayer) Stop() {
+	if p.cancel != nil {
+		p.cancel()
+		p.cancel = nil
+	}
+}
+
+func (p *StreamPlayer) download(it sdk.PlaylistFile) {
+	wg := &sync.WaitGroup{}
+	statusBar := &StatusBar{wg: wg}
+	wg.Add(1)
+
+	fileName := it.Name
+	localPath := filepath.Join(p.allocationID, fileName)
+
+	downloader, err := sdk.CreateDownloader(p.allocationID, localPath, it.Path,
+		sdk.WithAllocation(p.allocationObj),
+		sdk.WithAuthticket(p.authTicket, p.lookupHash))
+
+	if err != nil {
+		PrintError(err.Error())
+		return
+	}
+
+	defer sys.Files.Remove(localPath) //nolint
+
+	PrintInfo("playlist: downloading [", it.Path, "]")
+	err = downloader.Start(statusBar)
+
+	if err == nil {
+		wg.Wait()
+	} else {
+		PrintError("Download failed.", err.Error())
+		return
+	}
+	if !statusBar.success {
+		PrintError("Download failed: unknown error")
+		return
+	}
+
+	PrintInfo("playlist: downloaded [", it.Path, "]")
+	fs, _ := sys.Files.Open(localPath)
+
+	mf, _ := fs.(*sys.MemFile)
+
+	//AppendVideo(mf.Buffer.Bytes())
+
+	p.downloadedQueue <- mf.Buffer.Bytes()
+
+}
+
+func (p *StreamPlayer) startDownload() {
+	for {
+		select {
+		case <-p.ctx.Done():
+			PrintInfo("playlist: download is cancelled")
+			close(p.todoQueue)
+			close(p.reloadQueue)
+			return
+		case it := <-p.todoQueue:
+			p.download(it)
+
+			go p.nextTodo()
+		}
+
+	}
+}
+
+func (p *StreamPlayer) nextTodo() {
+	if len(p.todoQueue) < p.prefetchQty {
+		p.reloadQueue <- p.latestTodo
```

Comment:
This is a bit confusing as the reloadQueue receiver seems don't care what the value it received in the `reloadList()` method below, it would always use the p.latestTodo.

Reply:
rewritten it by j`s.setInterval`